File: src/fantasy_value/valuation/adapter.py
from __future__ import annotations

from typing import Iterable

from fantasy_value.projections.models import PlayerProjection
from fantasy_value.scoring.calculator import score_projection
from fantasy_value.scoring.models import ScoringConfig
from fantasy_value.valuation.models import ScoredPlayer


def to_scored_players(projections: Iterable[PlayerProjection], config: ScoringConfig) -> list[ScoredPlayer]:
    """The only bridge between real projection/scoring data and the
    valuation algorithms -- everything downstream of this stays
    schema-independent (the frontend re-implements it in TypeScript)."""
    scored: list[ScoredPlayer] = []
    for projection in projections:
        fantasy_positions = projection.fantasy_positions
        position = projection.position if projection.position in fantasy_positions else (
            fantasy_positions[0] if fantasy_positions else projection.position
        )
        if not position:
            raise ValueError(f"player {projection.player_id!r} has no position or fantasy_positions")

        points = score_projection(projection, config)
        scored.append(
            ScoredPlayer(
                player_id=projection.player_id,
                position=position,
                points=points,
                eligible_positions=frozenset(fantasy_positions) | {position},
            )
        )
    return scored

File: frontend/src/lib/valuation/adapter.ts
// Port of src/fantasy_value/valuation/adapter.py -- the only bridge between
// real projection/scoring data and the valuation algorithms. Everything
// downstream stays schema-independent.

import { scoreProjection } from '../scoring/calculator'
import type { ScoringConfig } from '../scoring/types'
import type { PlayerProjection } from '../projections/types'
import { makeScoredPlayer, type ScoredPlayer } from './models'

export function toScoredPlayers(
  projections: Iterable<PlayerProjection>,
  config: ScoringConfig,
): ScoredPlayer[] {
  const scored: ScoredPlayer[] = []
  for (const projection of projections) {
    const fantasyPositions = projection.fantasy_positions
    const position =
      projection.position && fantasyPositions.includes(projection.position)
        ? projection.position
        : (fantasyPositions[0] ?? projection.position)

    if (!position) {
      throw new Error(`player ${JSON.stringify(projection.player_id)} has no position or fantasy_positions`)
    }

    const points = scoreProjection(projection, config)
    scored.push(
      makeScoredPlayer({
        player_id: projection.player_id,
        position,
        points,
        eligible_positions: fantasyPositions,
      }),
    )
  }
  return scored
}
